Speed up startup and reload times

jvican opened the pull request:
These changes are optimizations that affect the following sbt subsytems:

1. Sbt parser.
2. Plugin discovery.

A concise explanation for every change is provided in the commit message.  In
short, this PR makes sure that we don't load the whole Scala reflect universe
to check whether a plugin has a `autoImport` member or to parse a sbt file.

In informal measurements, these changes affect around ~12-14% of the whole
startup time. These are the results before:

```
jvican in /data/rw/code/scala/scala
> $ time sbt "exit"
[info] Loading global plugins from /home/jvican/.sbt/0.13/plugins
[info] Loading project definition from /data/rw/code/scala/scala/project/project
[info] Loading project definition from /data/rw/code/scala/scala/project
[info] Resolving key references (11310 settings) ...
[info] *** Welcome to the sbt build definition for Scala! ***
[info] Check README.md for more information.
sbt -mem 2048 "exit"  29.66s user 0.56s system 297% cpu 10.171 total

jvican in /data/rw/code/scala/scala
> $ time sbt "exit"
[info] Loading global plugins from /home/jvican/.sbt/0.13/plugins
[info] Loading project definition from /data/rw/code/scala/scala/project/project
[info] Loading project definition from /data/rw/code/scala/scala/project
[info] Resolving key references (11310 settings) ...
[info] *** Welcome to the sbt build definition for Scala! ***
[info] Check README.md for more information.
sbt -mem 2048 "exit"  30.52s user 0.46s system 299% cpu 10.337 total

jvican in /data/rw/code/scala/scala
> $ time sbt "exit"
[info] Loading global plugins from /home/jvican/.sbt/0.13/plugins
[info] Loading project definition from /data/rw/code/scala/scala/project/project
[info] Loading project definition from /data/rw/code/scala/scala/project
[info] Resolving key references (11310 settings) ...
[info] *** Welcome to the sbt build definition for Scala! ***
[info] Check README.md for more information.
sbt -mem 2048 "exit"  28.85s user 0.43s system 293% cpu 9.988 total
```

And after:

```
jvican in /data/rw/code/scala/scala
> $ time sbt "exit"
[warn] Executing in batch mode.
[warn]   For better performance, hit [ENTER] to switch to interactive mode, or
[warn]   consider launching sbt without any commands, or explicitly passing 'shell'
[info] Loading global plugins from /home/jvican/.sbt/0.13/plugins
[info] Loading project definition from /data/rw/code/scala/scala/project/project
[info] Loading project definition from /data/rw/code/scala/scala/project
[info] Resolving key references (11355 settings) ...
[info] *** Welcome to the sbt build definition for Scala! ***
[info] Check README.md for more information.
sbt -mem 2048 "exit"  24.83s user 0.51s system 286% cpu 8.831 total

jvican in /data/rw/code/scala/scala
> $ time sbt "exit"
[warn] Executing in batch mode.
[warn]   For better performance, hit [ENTER] to switch to interactive mode, or
[warn]   consider launching sbt without any commands, or explicitly passing 'shell'
[info] Loading global plugins from /home/jvican/.sbt/0.13/plugins
[info] Loading project definition from /data/rw/code/scala/scala/project/project
[info] Loading project definition from /data/rw/code/scala/scala/project
[info] Resolving key references (11355 settings) ...
[info] *** Welcome to the sbt build definition for Scala! ***
[info] Check README.md for more information.
sbt -mem 2048 "exit"  24.46s user 0.46s system 287% cpu 8.674 total

jvican in /data/rw/code/scala/scala
> $ time sbt "exit"
[warn] Executing in batch mode.
[warn]   For better performance, hit [ENTER] to switch to interactive mode, or
[warn]   consider launching sbt without any commands, or explicitly passing 'shell'
[info] Loading global plugins from /home/jvican/.sbt/0.13/plugins
[info] Loading project definition from /data/rw/code/scala/scala/project/project
[info] Loading project definition from /data/rw/code/scala/scala/project
[info] Resolving key references (11355 settings) ...
[info] *** Welcome to the sbt build definition for Scala! ***
[info] Check README.md for more information.
sbt -mem 2048 "exit"  25.12s user 0.45s system 290% cpu 8.802 total
```

jvican committed:
Don't use runtime universe to discover autoImport

The previous implementation was using the Scala runtime universe to
check whether a plugin had or not an `autoImport` member. This is a bad
idea for the following reasons:

* The first time you use it, you class load the whole Scalac compiler
  universe. Not efficient. Measurements say this is about a second.
* There is a small overhead of going through the reflection API.

There exists a better approach that consists in checking if `autoImport`
exists with pure Java reflection. Since the class is already class
loaded, we check for:

* A class file named after the plugin FQN that includes `autoImport$` at
  the end, which means that an object named `autoImport` exists.
* A field in the plugin class that is named `autoImport`.

This complies with the plugin sbt specification:

http://www.scala-sbt.org/1.0/docs/Plugins.html#Controlling+the+import+with+autoImport

jvican committed:
Reuse the same global instance for parsing

The previous implementation was instantiating a toolbox to parse every
time it parsed a sbt file (and even recursively!!!).

This is inefficient and translates to instantiating a `ReflectGlobal`
every time we want to parse something.

This commit takes another approach:

1. It removes the dependency on `ReflectGlobal`.
2. It reuses the same `Global` and `Run` instances for parsing.

This is an efficient as it can get without doing a whole overhaul of it.
I think that in the future we may want to reimplement it to avoid the
recursive parsing to work around Scalac's bug.

jvican committed:
Avoid the use of `synchronized` while parsing

Previous commit used `synchronized` to ensure that the global reporter
was not reporting errors from other parsing sessions. Theoretically,
though, sbt could invoke parsing in parallel, so it's better to ensure
we remove the `synchronized` block, which could also be preventing some
JVM optimizations.

The following commit solves the issue by introducing a reporter id.

A reporter id is a unique identifier that is mapped to a reporter. Every
parsing session gets its own identifier, which then is reused for
recursive parsing. Error reports between recursive parses cannot collide
because the reporter is cleaned in `parse`.

eed3si9n commented on main/src/main/scala/sbt/internals/parser/SbtParser.scala:
```diff
@@ -6,14 +6,100 @@ import java.io.File
+    settings.outputDirs.setSingleOutput(new VirtualDirectory("(memory)", None))
+
+    // Mixing RangePositions is necessary, otherwise global ignores -Yrangepos
+    val global = new Global(settings, parserReporter) with RangePositions
```
Isn't the purpose of the toolbox to provide some abstraction instead of strongly coupling the implementation of sbt with Scala compiler? It is possible to reuse the toolbox, or would that incur initialization penalty?

/cc @adriaanm, @retronym

jvican replied:
This approach is better than reusing a toolbox, that relies on the `ReflectGlobal`. I explain more in the commit message.
  
This part of sbt already has access to `Global` since it uses it in `Eval`. There's no reason to use the toolbox, since the toolbox is meant for applications that don't have scala compiler as dependencies.

dwijnand commented on main/src/main/scala/sbt/Plugins.scala:
```diff
@@ -341,17 +341,15 @@ ${listConflicts(conflicting)}""")
+    val pluginClazz = ap.getClass
+    val autoImportExists = catching(classOf[ClassNotFoundException]).opt {
+      val autoImportClassName = s"${pluginClazz.getName}autoImport$$"
+      val fields = pluginClazz.getDeclaredFields
```
I wonder if "getMethod" is better than searching the fields.

jvican replied:
It is, I'll replace.

dwijnand replied:
When you do, mind the NoSuchMethodExceptions.

dwijnand commented on main/src/main/scala/sbt/Plugins.scala:
```diff
@@ -341,17 +341,15 @@ ${listConflicts(conflicting)}""")
+      val autoImportClassName = s"${pluginClazz.getName}autoImport$$"
+      val fields = pluginClazz.getDeclaredFields
+      fields.exists(_.getName == "autoImport") ||
+        {Class.forName(autoImportClassName, false, loader); true}
```
I think you need to handle ClassNotFoundException: [Class#forName](http://docs.oracle.com/javase/6/docs/api/java/lang/Class.html#forName(java.lang.String,%20boolean,%20java.lang.ClassLoader))

eed3si9n replied:
Isn't that handled with `catching(classOf[ClassNotFoundException]).opt` a few lines above?

jvican replied:
Indeed.

dwijnand replied:
You're right. I'm blind. But personally I'd try and wrap things as small as possible.

dwijnand commented on main/src/main/scala/sbt/internals/parser/SbtParser.scala:
```diff
@@ -6,14 +6,100 @@ import java.io.File
+        val errorMessage = seq.mkString(EOL)
+        val error: String =
+          if (errorMessage.contains(XML_ERROR))
+            s"$errorMessage\n${SbtParser.XmlErrorMessage}"
```
You can drop the `SbtParser.` here (and thus the braces)

dwijnand commented on main/src/main/scala/sbt/internals/parser/SbtParser.scala:
```diff
@@ -3,17 +3,144 @@ package internals
+    val run = new global.Run
+    // Add required dummy unit for initialization...
+    val initFile = new BatchSourceFile("<wrapper-init>", "")
+    val _ = new global.CompilationUnit(initFile)
```
is the LHS val necessary? Do we depend on side-effects from calling the constructor of CompilationUnit?

jvican replied:
Yes, black magic of the compiler...

dwijnand replied:
yes to both?? or just the 2nd question?

jvican replied:
Just the last one, the reason I use `val _` is to make it explicit that the line is really required and it's not any dead code (some IDEs may report otherwise).

dwijnand replied:
Some IDEs got to get with the system. We _side-effect_ in _constructors_ of course!

dwijnand reviewed (changes requested):
LGTM, apart from the ClassNotFoundException part.

The rest are thoughts, questions and minor points/issues.

dwijnand commented:
2 more things: (1) I restarted the failure on Travis CI, and (2) nice work!

retronym commented:
I'd suggest to drop the last commit for now. Parsing is really quite cheap once you've got the compiler initialized, so I don't think future proofing this code so as not to be a point of contention for a hypothetical multi-threaded project load is worth.

retronym commented:
I'd also suggest starting with an initial commit that reuses the `ToolBox` based parser. If for some reason, we have to revert the change to use `Global` directly (hypothetically this might happen if the compiler API changes in a future compiler that SBT upgrades to and this code is struggling to fina a maintainer), we'll still have the main performance improvement.

retronym commented on main/src/main/scala/sbt/Plugins.scala:
```diff
@@ -341,17 +341,15 @@ ${listConflicts(conflicting)}""")
+    val pluginClazz = ap.getClass
+    val autoImportExists = catching(classOf[ClassNotFoundException]).opt {
+      val autoImportClassName = s"${pluginClazz.getName}autoImport$$"
+      val fields = pluginClazz.getDeclaredFields
```
This should walk up the superclass chain in case the `AutoPlugin` inherits the `autoImport` member.

We no longer can check whether or not `autoImport` is a stable term (`val` or `object`), but I think it is fine to document that plugin authors must not declare `var autoImport`.

jvican replied:
You're right, this should also be done for super classes. I actually thought this was not possible in the current plugin architecture.

jvican replied:
I've tried this out, but it's not as simple as it seems since the tests check that the current sbt behaviour is exactly the same as the one before, and there are some mismatches in the way both parsers handle positions and new lines. I'm looking into this problem and hope to reconcile them, but we should probably stage this to another PR.

retronym commented on main/src/main/scala/sbt/internals/parser/SbtParser.scala:
```diff
@@ -130,7 +179,7 @@ private[sbt] case class SbtParser(file: File, lines: Seq[String]) extends Parsed
+      val statement = util.Try(parse(originalStatement, fileName)) match {
         case util.Failure(th) =>
           val missingText = findMissingText(content, t.pos.end, t.pos.line, fileName, th)
           originalStatement + missingText
```
I know this triple-parsing rigmarole bugs you, so I thought I'd leave you with the idea to ignore the end position of parsed statements, and instead  use the start position of the subsequent statement (or the end position of the enclosing tree, if there is no subsequent statement) - 1.

jvican replied:
This is such a great idea. Thank you.

jvican commented:
> I'd also suggest starting with an initial commit that reuses the ToolBox based parser. If for some reason, we have to revert the change to use Global directly (hypothetically this might happen if the compiler API changes in a future compiler that SBT upgrades to and this code is struggling to fina a maintainer), we'll still have the main performance improvement.
  
The problem is that toolbox is not thread-safe. You cannot use it in the same way I'm using it now because reporter errors will clash with each other. This is the reason why I'm doing all the magic with the `reporterId`s. I actually think that we may want to parse sbt files in parallel (if sbt is not already doing this). I heard somewhere that scalac was parsing in parallel too.
    
> Parsing is really quite cheap once you've got the compiler initialized

Isn't there an overhead of using `ReflectGlobal` over just `Global`? The current approach in this commit, as I see it, avoids class loading some classes.
  
True, parsing is quite cheap. The reason I went for this solution though is because I gained 400ms. I'll repeat my measurements with the global toolbox, but we still have the issue that is not thread-safe and we cannot instrument it.
  
However, a bigger benefit with this approach is that we can reuse the parsed trees from this concrete parser and feed them into eval, instead of parsing the whole compilation unit again. This will reduce quite some code and ease maintenance. Potentially, we could even reuse the global of eval for parsing, avoiding the initialization of the extra global, which I believe it may be non-negligible. WDYT?

retronym commented:
The lack of thread safety is moot if we live with synchronisation. I
believe will still be a big win by avoiding unneeded init cost

I'm in favour of the end result of using global directly; but I think it's
worth an extra commit to make that a separate step.

jvican commented:
@retronym So what you propose is that I include a commit before my changes to parser that uses a global toolbox instance and is synchronized, no?
  
I'm happy to do that, I have that commit lying around my computer somewhere.

retronym commented:
👍

jvican commented:
I'm closing this because I'm reopening against 1.0.

jvican committed:
Add global toolbox to parse sbt files

This change was proposed by Jason in case that the new parsing mechanism
implemented later on has to be reverted. This change provides a good
baseline, but it's far from ideal with regard to readability of the
parser and performance.

jvican commented:
Hello, Jason's changes have been applied in https://github.com/sbt/sbt/pull/3115, which is rebased to latest master 1.0.
